components/org.wso2.carbon.identity.governance/src/main/java/org/wso2/carbon/identity/governance/listener/ClaimValueEncryptionListener.java

```diff
@@ -0,0 +1,239 @@
+            return orderId;
+        }
+        return 94;
+    }
+
+    @Override
+    public boolean doPreAddUserWithID(String userID, Object credential, String[] roleList, Map<String, String> claims,
+                                      String profile, UserStoreManager userStoreManager) throws UserStoreException {
+
+        try {
+            if (!isEnable() || userStoreManager == null) {
+                return true;
+            }
+
+            for (Map.Entry<String, String> entry : claims.entrySet()) {
+                String claimURI = entry.getKey();
+                if (checkEnableEncryption(claimURI, userStoreManager)) {
+                    String claimValue = entry.getValue();
+                    try {
+                        claimValue = encryptClaimValue(claimValue);
+                    } catch (CryptoException e) {
+                        LOG.error("Error occurred while encrypting claim value", e);
+                    }
+                    claims.put(claimURI, claimValue);
+                }
+            }
+            return true;
+        } catch (org.wso2.carbon.user.api.UserStoreException e) {
+            throw new UserStoreException("Error while retrieving tenant ID", e);
+        }
+    }
+
+    @Override
+    public boolean doPreSetUserClaimValues(String userName, Map<String, String> claims, String
+            profileName, UserStoreManager userStoreManager) throws UserStoreException {
+
+        if (!isEnable() || userStoreManager == null) {
+            return true;
+        }
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Pre set claims is called in ClaimValueEncryptionListener");
+        }
+
+        for (Map.Entry<String, String> entry : claims.entrySet()) {
+            String claimURI = entry.getKey();
+            if (checkEnableEncryption(claimURI, userStoreManager)) {
+                String claimValue = entry.getValue();
+                try {
+                    claimValue = encryptClaimValue(claimValue);
+                } catch (CryptoException e) {
+                    LOG.error("Error occurred while encrypting claim value", e);
+                }
+                claims.put(claimURI, claimValue);
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean doPreSetUserClaimValue(String userName, String claimURI, String claimValue, String profileName,
+                                          UserStoreManager userStoreManager) throws UserStoreException {
+
+        if (!isEnable() || userStoreManager == null) {
+            return true;
+        }
+        if (checkEnableEncryption(claimURI, userStoreManager)) {
+            IdentityUtil.threadLocalProperties.get().remove(CLAIM_URI);
+            IdentityUtil.threadLocalProperties.get().remove(CLAIM_VALUE);
+
+            IdentityUtil.threadLocalProperties.get().put(CLAIM_URI, claimURI);
+            IdentityUtil.threadLocalProperties.get().put(CLAIM_VALUE, claimValue);
```

Comment:
why is this just setting the incoming value as it is?

Reply:
Those values will be used in the post listener. As in this method claim value is passed as a `String` we cannot modify the claim value(encrypt) and propagate it to the next listener.